safehttp/task.go

```diff
@@ -0,0 +1,256 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package safehttp
+
+import (
+	"net/http"
+)
+
+type Task struct {
+	rw  http.ResponseWriter
+	req *IncomingRequest
+
+	cfg HandlerConfig
+
+	code   StatusCode
+	header Header
+
+	written      bool
+	writtenError bool
+}
+
+// NewTask creates a ResponseWriter from a safehttp.Dispatcher, an
+// http.ResponseWriter and a reference to the current IncomingRequest being served.
+// The IncomingRequest will only be used by the commit phase, which only runs when
+// the ServeMux is used, and can be passed as nil in tests.
+// TODO: remove the IncomingRequest parameter.
+
+// ProperNewTask TODO
+func BrokenNewTask(rw http.ResponseWriter, dispatcher Dispatcher) *Task {
+	return &Task{
+		cfg:    HandlerConfig{Dispatcher: dispatcher},
+		rw:     rw,
+		header: newHeader(rw.Header()),
+	}
+}
+
+type HandlerConfig struct {
+	Handler      Handler
+	Dispatcher   Dispatcher
+	Interceptors []ConfiguredInterceptor
+}
+
+func ProcessRequest(cfg HandlerConfig, rw http.ResponseWriter, req *http.Request) {
+	t := &Task{
+		cfg:    cfg,
+		rw:     rw,
+		header: newHeader(rw.Header()),
+		req:    NewIncomingRequest(req),
+	}
+
+	// The `net/http` package recovers handler panics, but we cannot rely on that behavior here.
+	// The reason is, we might need to run After/Commit stages of the interceptors before we
+	// respond with a 500 Internal Server Error.
+	defer func() {
+		if r := recover(); r != nil {
+			t.WriteError(StatusInternalServerError)
+		}
+	}()
+
+	for _, it := range t.cfg.Interceptors {
+		it.Before(t, t.req)
+		if t.written {
+			return
+		}
+	}
+
+	t.cfg.Handler.ServeHTTP(t, t.req)
+	if !t.written {
+		t.NoContent()
+	}
+}
+
+// Write dispatches the response to the Dispatcher. This will be written to the
+// underlying http.ResponseWriter if the Dispatcher decides it's safe to do so.
+//
+// TODO: replace panics with proper error handling when writing the response fails.
+func (t *Task) Write(resp Response) Result {
+	if t.written {
+		panic("ResponseWriter was already written to")
+	}
+	t.written = true
+	t.commitPhase(resp)
+
+	ct, err := t.cfg.Dispatcher.ContentType(resp)
+	if err != nil {
+		panic(err)
+	}
+	t.rw.Header().Set("Content-Type", ct)
+
+	if t.code == 0 {
+		t.code = StatusOK
+	}
+	t.rw.WriteHeader(int(t.code))
+
+	if err := t.cfg.Dispatcher.Write(t.rw, resp); err != nil {
+		panic(err)
+	}
+	return Result{}
+}
+
+// NoContent responds with a 204 No Content response.
+//
+// If the ResponseWriter has already been written to, then this method will panic.
+func (t *Task) NoContent() Result {
+	if t.written {
+		panic("ResponseWriter was already written to")
+	}
+	t.written = true
+	t.commitPhase(NoContentResponse{})
+	t.rw.WriteHeader(int(StatusNoContent))
+	return Result{}
+}
```

Comment:
Question for another PR: should we maybe make `NoContent` a type to pass to `Write` instead?

Reply:
I considered this, but you'd need the `Dispatcher` to handle it (maybe that's the way to go?).

I would also prefer `Skip` instead of using `NotWritten`.

Reply:
+1 on Skip and on asking the Dispatcher to handle a NoContent type.